core/frontend/src/components/vehiclesetup/viewers/GenericViewer.vue

```diff
@@ -0,0 +1,404 @@
+    this.reloadAnnotations()
+  },
+  methods: {
+    async reloadAnnotations() {
+      const json = await models(`./${this.vehicle_folder}/${this.frame_name}.json`)
+      if (json) {
+        this.annotations = json.annotations ?? {}
+      }
+    },
+    redraw() {
+      this.setAlphas(1.0)
+      this.hideIrrelevantParts()
+      this.forceRefreshAnnotations()
+    },
+    async checkModelOverrides() {
+      while (!this.vehicle_type || !this.frame_name) {
+        await sleep(100)
+      }
+      const master_override = '/userdata/modeloverrides/ALL.glb'
+      const vehicle_override = `/userdata/modeloverrides/${this.vehicle_folder}/${this.frame_name}.glb`
+      try {
+        await axios.head(master_override)
+        return master_override
+      } catch {
+        console.log(`master override model not found at ${master_override}`)
+      }
+      try {
+        await axios.head(vehicle_override)
+        return vehicle_override
+      } catch {
+        console.log(`vehicle override model not found at ${vehicle_override}`)
+      }
+      return undefined
+    },
+    async loadAnottationsOverride(): Promise<Dictionary<HotspotConfiguration>> {
+      if (!this.model_override_path) {
+        return {}
+      }
+      const candidate_path = this.model_override_path?.replace('glb', 'json')
+      const response = await axios.get(candidate_path)
+      return response.data?.annotations ?? {}
+    },
+    setAlphas(new_color: number, text = ''): void {
+      const lower_text = text.toLowerCase()
+      // eslint-disable-next-line no-extra-parens
+      const materials = (this.$refs.modelviewer as ModelViewerElement).model?.materials ?? []
+      const affected_materials = materials.filter(
+        (material) => material.name.toLowerCase().includes(lower_text),
+      )
+      for (const material of affected_materials) {
+        // store default alphas and do not allow going above it.
+        if (!(material.name in this.default_alphas)) {
+          // eslint-disable-next-line prefer-destructuring
+          this.default_alphas[material.name] = material.pbrMetallicRoughness.baseColorFactor[3]
+        }
+        const color = material.pbrMetallicRoughness.baseColorFactor
+        color[3] = Math.min(new_color, this.default_alphas[material.name])
+        material.setAlphaMode(color[3] < 1.0 ? 'BLEND' : 'OPAQUE')
+        material.pbrMetallicRoughness.setBaseColorFactor(color)
+      }
+    },
+    hideIrrelevantParts(): void {
+      if (!this.gripper_is_present) {
+        this.setAlphas(0.0, 'gripper')
+      }
+      if (!this.lights1_are_present) {
+        this.setAlphas(0.0, 'lights 1')
+      }
+      if (!this.lights2_are_present) {
+        this.setAlphas(0.0, 'lights 2')
+      }
+      if (!this.ping1D_is_present) {
+        this.setAlphas(0.0, 'ping1d')
+      }
+      if (!this.ping360_is_present) {
+        this.setAlphas(0.0, 'ping360')
+      }
+    },
```

Comment:
Is it possible for us to handle this via the json ?

Reply:
i cant think of way to do this..

Reply:
Could we enforce the name needing to be the bit before the `is_present`? The names are already close to that being the case, and then the json could just list the properties of interest, and the `.vue` file could just check something like

```ts
if (!this.{name}_is_present) {
    this.setAlphas(0.0, name)
}
```

In Python that would be something like
```py
if not getattr(self, f'{name}_is_present', None):
    self.set_alphas(0.0, name)
```
Not sure what the typescript equivalent of `getattr` is.

Of course the `X_is_present` functions still need to be defined somewhere, but that structure would at least allow selecting predefined functions 🤷‍♂️

Reply:
I think we cant really rely on names, as the code gets minified

Reply:
yeah, but I don't see the point right now. if planes dont support the gripper, for example, they can just omit it from the 3d model.

Reply:
though we should probably improve the gripper detection